crates/sui-graphql-rpc/src/types/transaction_block.rs

```diff
@@ -297,102 +305,212 @@ impl TransactionBlock {
+        // bound. Similarly, decrement `before` by 1 so that we can uniformly select the
+        // `max_tx_sequence_number`. In other words, the range consists of all transactions from the
+        // smallest tx_sequence_number in lo_cp to the max tx_sequence_number in hi_cp.
+        let mut lo_cp = max_option!(
+            filter.after_checkpoint.map(|x| x.saturating_add(1)),
+            filter.at_checkpoint,
+            Some(0)
+        )
+        .unwrap(); // Safe to unwrap because of the `Some(0)` default
+
+        let mut hi_cp = min_option!(
+            filter.at_checkpoint,
+            filter.before_checkpoint.map(|x| x.saturating_sub(1)),
+            Some(checkpoint_viewed_at)
+        )
+        .unwrap();
+
+        // If `within_checkpoints` is set, we need to adjust the lower and upper bounds. It is up to
+        // the caller of `TransactionBlock::paginate` to determine whether `within_checkpoints` is
+        // required.
+        if let Some(scan_limit) = within_checkpoints {
+            if page.is_from_front() {
+                hi_cp = std::cmp::min(hi_cp, lo_cp.saturating_add(scan_limit));
+            } else {
+                // Similar to the `first` case, `checkpoint_viewed_at` offers a fixed point to
+                // determine how many more checkpoints to scan.
+                lo_cp = std::cmp::max(lo_cp, hi_cp.saturating_sub(scan_limit));
+            }
+        }
 
-                        if let Some(o) = &filter.changed_object {
-                            let sub_query = tx_changed_objects::dsl::tx_changed_objects
-                                .select(tx_changed_objects::dsl::tx_sequence_number)
-                                .filter(tx_changed_objects::dsl::object_id.eq(o.into_vec()));
-                            query = query.filter(tx::dsl::tx_sequence_number.eq_any(sub_query));
-                        }
+        use checkpoints::dsl as cp;
+        use transactions::dsl as tx;
 
-                        if let Some(txs) = &filter.transaction_ids {
-                            let digests: Vec<_> = txs.iter().map(|d| d.to_vec()).collect();
-                            query = query.filter(tx::dsl::transaction_digest.eq_any(digests));
+        let (prev, next, transactions): (bool, bool, Vec<StoredTransaction>) = db
+            .execute_repeatable(move |conn| {
+
+                // The min or first `tx_sequence_number` of a checkpoint is the previous
+                // checkpoint's `network_total_transactions`. Because this refers to a historical
+                // checkpoint, if we yield a `None` result, we can return early.
+                let lo_tx = match lo_cp {
+                    0 => 0,
+                    _ => {
+                        let sequence_number: Option<i64> = conn
+                        .first(move || {
+                            cp::checkpoints
+                                .select(cp::network_total_transactions)
```

Comment:
These kinds of filters should be using the tx sequence number upper and lower bounds.